File: Programmers/올바른-괄호.py
def solution(s):
    answer = True
    stack = []
    count = 0
    
    for i in range(len(s)):
        if s[i] == '(':
            stack.append(s[i])
            count += 1 
        elif s[i] == ')':
            if len(stack) == 0:
                return False
            stack.pop()
            count -= 1

    if count != 0:
        answer = False

    return answer

File: java/inflearn/올바른괄호.java
import java.util.Scanner;
import java.util.Stack;


public class Main {

    public static String solution(String str){
        String answer = "YES";
        Stack<Character> stack = new Stack<>();

        for(char c : str.toCharArray()){
            if(c == '('){
                stack.push(c);
            } else{
                if(stack.isEmpty()){
                    return "NO";
                }
                stack.pop();
            }
        }
        if (!stack.isEmpty()){
            return "NO";
        }
        return answer;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.next();
        System.out.println(solution(str));
    }
}
